extensions/wikia/UserLogin/UserLoginSpecialController.class.php

```diff
@@ -589,39 +591,75 @@ public function mailPassword() {
-		} else if ( !$this->wg->Auth->allowPasswordChange() ) {
-			$this->result = 'error';
-			$this->msg = wfMessage( 'userlogin-error-resetpass_forbidden' )->escaped();
-		} else if ( $this->wg->User->isBlocked() ) {
-			$this->result = 'error';
-			$this->msg = wfMessage( 'userlogin-error-blocked-mailpassword' )->escaped();
+			$this->setErrorResponse( 'userlogin-error-noname' );
+			return;
+		}
+
+		if ( !$this->wg->Auth->allowPasswordChange() ) {
+			$this->setErrorResponse( 'userlogin-error-resetpass_forbidden' );
+			return;
+		}
+
+		if ( $this->wg->User->isBlocked() ) {
+			$this->setErrorResponse( 'userlogin-error-blocked-mailpassword' );
+			return;
+		}
+
+		$user = User::newFromName( $loginForm->mUsername );
+		if ( !$user instanceof User ) {
+			$this->setErrorResponse( 'userlogin-error-noname' );
+			return;
+		}
+
+		if ( $user->getID() == 0 ) {
+			$this->setErrorResponse( 'userlogin-error-nosuchuser' );
+			return;
+		}
+
+		if ( $user->isPasswordReminderThrottled() ) {
+			$throttleTTL = round( $this->wg->PasswordReminderResendTime, 3 );
+			$this->setErrorResponse( 'userlogin-error-throttled-mailpassword', $throttleTTL );
+			return;
+		}
+
+		/// Get a temporary password
+		$userService = new \UserService();
+		$tempPass = $userService->resetPassword( $user );
+
+		$resp = F::app()->sendRequest( 'Email\Controller\ForgotPassword', 'handle', [
+			'targetUser' => $user,
+			'tempPass' => $tempPass,
+		] );
+
+		$data = $resp->getData();
+		if ( !empty( $data['result'] ) && $data['result'] == 'ok' ) {
+			$this->setSuccessResponse( 'userlogin-password-email-sent', $loginForm->mUsername );
 		} else {
-			$user = User::newFromName( $loginForm->mUsername );
-			if ( !$user instanceof User ) {
-				$this->result = 'error';
-				$this->msg = wfMessage( 'userlogin-error-noname' )->escaped();
-			} else if ( $user->getID() == 0 ) {
-				$this->result = 'error';
-				$this->msg = wfMessage( 'userlogin-error-nosuchuser' )->escaped();
-			} else if ( $user->isPasswordReminderThrottled() ) {
-				$this->result = 'error';
-				$this->msg = wfMessage( 'userlogin-error-throttled-mailpassword', round( $this->wg->PasswordReminderResendTime, 3 ) )->escaped();
-			} else {
-				$emailTextTemplate = $this->app->renderView( "UserLogin", "GeneralMail", array( 'language' => $user->getOption( 'language' ), 'type' => 'password-email' ) );
-				$result = $loginForm->mailPasswordInternal( $user, true, 'userlogin-password-email-subject', 'userlogin-password-email-body', $emailTextTemplate );
-				if ( !$result->isGood() ) {
-					$this->result = 'error';
-					$this->msg = wfMessage( 'userlogin-error-mail-error', $result->getMessage() )->parse();
-				} else {
-					$this->result = 'ok';
-					$this->msg = wfMessage( 'userlogin-password-email-sent', $loginForm->mUsername )->escaped();
-				}
-			}
+			$this->setParsedErrorResponse( 'userlogin-error-mail-error' );
 		}
 	}
 
+	private function setErrorResponse( ...$params ) {
```

Comment:
It would kind of complicate things.  If I specified ```$key``` here, it would no longer be in $params.  At some point I would have to add it back to $params without ever being used directly.